src/Entity/Storage/ConstraintEntityStorage.php

```diff
@@ -0,0 +1,164 @@
+<?php
+
+namespace Drupal\password_enhancements\Entity\Storage;
+
+use Drupal\Component\Uuid\UuidInterface;
+use Drupal\Core\Cache\MemoryCache\MemoryCacheInterface;
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\Core\Config\Entity\ConfigEntityStorage;
+use Drupal\Core\Config\Entity\ConfigEntityStorageInterface;
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Entity\EntityTypeInterface;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Language\LanguageManagerInterface;
+use Drupal\password_enhancements\Entity\Constraint;
+use Drupal\password_enhancements\Logger\Logger;
+use Drupal\password_enhancements\PasswordConstraintPluginManager;
+use Exception;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Entity storage for password constraint config entity.
+ */
+class ConstraintEntityStorage extends ConfigEntityStorage implements ConstraintEntityStorageInterface {
+
+  /**
+   * Entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * Logger.
+   *
+   * @var \Drupal\password_enhancements\Logger\Logger
+   */
+  protected $logger;
+
+  /**
+   * Password constraint plugin manager.
+   *
+   * @var \Drupal\password_enhancements\PasswordConstraintPluginManager
+   */
+  protected $passwordConstraintPluginManager;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(EntityTypeInterface $entity_type, ConfigFactoryInterface $config_factory, UuidInterface $uuid_service, LanguageManagerInterface $language_manager, PasswordConstraintPluginManager $password_constraint_plugin_manager, MemoryCacheInterface $memory_cache, EntityTypeManagerInterface $entity_type_manager, Logger $logger) {
+    parent::__construct($entity_type, $config_factory, $uuid_service, $language_manager, $memory_cache);
+
+    $this->entityTypeManager = $entity_type_manager;
+    $this->logger = $logger;
+    $this->passwordConstraintPluginManager = $password_constraint_plugin_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function createInstance(ContainerInterface $container, EntityTypeInterface $entity_type): ConfigEntityStorageInterface {
+    return new static(
+      $entity_type,
+      $container->get('config.factory'),
+      $container->get('uuid'),
+      $container->get('language_manager'),
+      $container->get('plugin.manager.password_constraint'),
+      $container->get('entity.memory_cache'),
+      $container->get('entity_type.manager'),
+      $container->get('logger.password_enhancements')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function save(EntityInterface $entity): int {
+    // Generate new ID for the entity based on the uniqueness of the selected
+    // plugin.
+    /** @var \Drupal\password_enhancements\Entity\ConstraintInterface $entity */
+    $type = $entity->getType();
+    if (empty($type)) {
```

Comment:
This is just a sanity check, it should never happen though.

What validation phase are you mentioning? If I either call Entity::save() or I'm calling the save() method through the storage this always gets called.

Maybe doPreSave() method? It wouldn't work if I would try to validate it in the constructor of the class, because the form always creates an empty entity first.

Reply:
Well, Drupal makes it really complicated to do this properly... If this would be content entity then probably the easiest way to convert properties to basefields because you can attach validation handlers to base fields. But this is a config entity.

You have to ensure that Drupal validates the property's value before it sets it no matter how it does:
* via constructor
* via setter
* via magic setter
etc?

If you inherit from `EntityBase` Drupal does do any validation on a property when it sets in the constructor, just blindly set the value and if PHP does not yell then the value is set. 
(Magic setters also only do validation if they find a field on content entities, otherwise they just do the same as the constructor does: `\Drupal\Core\Entity\ContentEntityBase::__set()`).

What I did in the PHP API client to resolve this problem is that I made sure that even the constructor calls the setter of the property so I only had to add validation to the setter.

Reply:
You miss the point, the entity form has to create an empty Entity, if I would validate anywhere before saving the entity then the form would fail because it didn't got the right value.

Reply:
Well, I solved something similar before so let's get back to this as part of a follow-up task. #9